relayer/paths/path_processor.go

```diff
@@ -0,0 +1,499 @@
+package paths
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/cosmos/relayer/v2/relayer/chains/processor"
+	"github.com/cosmos/relayer/v2/relayer/ibc"
+	"github.com/cosmos/relayer/v2/relayer/provider"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type PathProcessor struct {
+	ctx context.Context
+	log *zap.Logger
+
+	pathEnd1 *PathEndRuntime
+	pathEnd2 *PathEndRuntime
+
+	processLock                    sync.Mutex
+	shouldProcessAgainOnceComplete bool
+	retryTimer                     *time.Timer
+}
+
+type PathEndRuntime struct {
+	info PathEnd
+
+	chainProcessor processor.ChainProcessor
+
+	// ibc Messages backlog
+	// retains minimal state for a given packet
+	// packet message history will be held onto until messages come in that signal the packet flow is complete
+	// when packet-flow-complete messages are handled (MsgAcknowledgement, MsgTimeout, or MsgTimeoutOnClose),
+	// the entire packet history for that sequence number will be cleared out
+	messages     map[string]map[uint64]provider.RelayerMessage
+	messagesLock sync.Mutex
+
+	// ibc Messages that are in progress sending
+	// Used so that cycles in the `process` method do not attempt to send the same messages twice
+	// also adds packet retry if a given packet fails to send after the threshold # blocks
+	inProgressMessageSends     map[string]map[uint64]ibc.InProgressSend
+	inProgressMessageSendsLock sync.Mutex
+
+	maxSequence     uint64
+	maxSequenceLock sync.Mutex
+}
+
+func (per *PathEndRuntime) updateMaxSequence(sequence uint64) {
+	if sequence > per.maxSequence {
+		per.maxSequence = sequence
+	}
+}
+
+func NewPathProcessor(ctx context.Context, log *zap.Logger, pathEnd1 PathEnd, pathEnd2 PathEnd) *PathProcessor {
+	return &PathProcessor{
+		ctx: ctx,
+		log: log,
+		pathEnd1: &PathEndRuntime{
+			info:                       pathEnd1,
+			messages:                   make(map[string]map[uint64]provider.RelayerMessage),
+			messagesLock:               sync.Mutex{},
```

Comment:
Zero values of mutexes are ready for use. Don't explicitly initialize them for struct fields.